broker-core/src/main/java/io/zeebe/broker/job/processor/JobInstanceStreamProcessor.java

```diff
@@ -65,16 +77,99 @@ public TypedStreamProcessor createStreamProcessor(TypedStreamEnvironment environ
         .onCommand(ValueType.JOB, JobIntent.TIME_OUT, new TimeOutJobProcessor())
         .onCommand(ValueType.JOB, JobIntent.UPDATE_RETRIES, new UpdateRetriesJobProcessor())
         .onCommand(ValueType.JOB, JobIntent.CANCEL, new CancelJobProcessor())
-        .withStateResource(jobIndex.getMap())
+        .withRocksDBController(dbController)
+        .withListener(this)
         .build();
   }
 
+  public StateSnapshotController createSnapshotController(StateStorage storage) {
+    return new StateSnapshotController(dbController, storage);
+  }
+
+  @Override
+  public void onOpen(TypedStreamProcessor streamProcessor) {
+    if (!dbController.isOpened()) {
+      throw new IllegalStateException("should have opened internal state database");
+    }
+
+    db = dbController.getDb();
+    recoverLatestJobKey(streamProcessor.getKeyGenerator());
+  }
+
+  @Override
+  public void onClose() {
+    dbController.close();
+    db = null;
+  }
+
+  private void recoverLatestJobKey(KeyGenerator keyGenerator) {
+    if (tryGet(LATEST_JOB_KEY_BUFFER, dbLongBuffer.array())) {
+      keyGenerator.setKey(dbLongBuffer.getLong(0));
+    }
+  }
+
+  private void putLatestJobKey(long key) {
+    dbLongBuffer.putLong(0, key);
+
+    try {
+      db.put(LATEST_JOB_KEY_BUFFER, dbLongBuffer.array());
+    } catch (RocksDBException e) {
+      LangUtil.rethrowUnchecked(e);
+    }
+  }
+
+  private void putJobState(long key, short state) {
```

Comment:
Fair enough. I thought it'd work well subclassing the `StateController`, at this point it might make sense then to rename it simply to State. What do you think? See `JobInstanceStateController`.

Reply:
It's ok but I think it is not necessary to extend the `StateContoller` because you only need the `RocksDB` instance.

Reply:
I agree, but I extended it because extending the class is the way to eventually customize the DB (creation options, DB type - e.g. if you need it to be a TransactionDB, or TtlDB, etc., and so forth), and I think that will be the way forward in the future. I'm also not sure I like having the DB instance exposed, actually, since the controller controls if it's opened/close, and having access to the DB instance outside means you could potentially close it somewhere else.